include/boost/geometry/policies/robustness/segment_ratio.hpp

```diff
@@ -135,11 +135,19 @@ struct possibly_collinear<Type, false>
 template <typename Type>
 class segment_ratio
 {
-public :
-    typedef Type numeric_type;
+    // Use the type (if it is already floating point), or a double
+    // (for integer types or user defined types).
+    // This is used for the approximation (a helper value)
+    // and for the edge value (0..1) (a helper function).
+    using floating_point_type = std::conditional_t
+        <
+            std::is_floating_point<Type>::value, Type, double
+        >;
```

Comment:
Since you moved it we may address its problem now. The problem is that user-defined FP type is going to be degraded to `double` so this typedef should rather be:
```
    using floating_point_type = std::conditional_t
        <
            std::is_integral<Type>::value, double, Type
        >;
```
After this change it would not work properly for user-defined integral types but AFAIU we assume that user-defined types are FP types.

In fact we should probably add traits allowing the users to specify how the library should handle user-defined numeric types but that's outside of the scope of this PR.

Reply:
Thanks for this and for your other remarks, I'm processing them.

One thing: this proposed change does not work for `boost::rational` because it will set floating_point_type to rational. 

Your suggestion (`add traits allowing ...`)  is very good and I will use that approach (not yet making it a generic traits though)

Reply:
Or actually it will be generic. We also have `fp_coordinate_type` and `promote_floating_point` which I will combine.

Reply:
Ok, though I think that they should be added consistently in the whole library which is outside of the scope of this PR. I'd be fine with a simple fix here.

Reply:
Agreed, I've something but it's growing. For now I'll apply a simple fix.

Reply:
:heavy_check_mark: